pkg/k8s/resource/scheme.go

```diff
@@ -0,0 +1,107 @@
+// SPDX-License-Identifier: Apache-2.0
+// Copyright Authors of Cilium
+
+package resource
+
+import (
+	"fmt"
+
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/runtime"
+	"k8s.io/apimachinery/pkg/runtime/schema"
+	utilruntime "k8s.io/apimachinery/pkg/util/runtime"
+
+	cilium_api_v2 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2"
+	cilium_api_v2alpha1 "github.com/cilium/cilium/pkg/k8s/apis/cilium.io/v2alpha1"
+	corev1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/core/v1"
+	discoveryv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/discovery/v1"
+	discoveryv1beta1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/discovery/v1beta1"
+	networkingv1 "github.com/cilium/cilium/pkg/k8s/slim/k8s/api/networking/v1"
+)
+
+var scheme = runtime.NewScheme()
+
+var localSchemeBuilder = runtime.SchemeBuilder{
+	corev1.AddToScheme,
+	discoveryv1beta1.AddToScheme,
+	discoveryv1.AddToScheme,
+	networkingv1.AddToScheme,
+	cilium_api_v2.AddToScheme,
+	cilium_api_v2alpha1.AddToScheme,
+}
+
+var AddToScheme = localSchemeBuilder.AddToScheme
+
+func init() {
+	utilruntime.Must(AddToScheme(scheme))
+}
+
+// referenced from https://github.com/kubernetes-sigs/controller-runtime/blob/93cd3cda0fdddc1f703634a881097c246dcabbd3/pkg/client/apiutil/apimachinery.go#L113
+// We are doing it this way to avoid the direct dependency on controller-runtime, otherwise conflicts in metric registration will occur
+// controller-runtime automatically registers metrics such as k8s rest client and workqueue metrics on its registry,
+// and cilium-agent also registers the same metrics on its registry. Those metrics registrations are exclusive.
+func gvkForObject(obj runtime.Object) (schema.GroupVersionKind, error) {
```

Comment:
Where does [apiutil](https://github.com/kubernetes-sigs/controller-runtime/tree/93cd3cda0fdddc1f703634a881097c246dcabbd3/pkg/client/apiutil) automatically register its metrics?

Reply:
On second thoughts, depending on `sigs.k8s.io/controller-runtime/pkg/client/apiutil` doesn't cause the metrics conflict because the workqueue metrics are registered in the different package
`sigs.k8s.io/controller-runtime/pkg/metrics`. So we can drop the GetGVK function here and use the original function. Thank you for your comment!

https://github.com/kubernetes-sigs/controller-runtime/blob/f30e11d97fd19564e71b4f672efcdc0ecad3a8d1/pkg/metrics/workqueue.go#L90-L100